torch/_refs/nn/functional/__init__.py

```diff
@@ -127,6 +128,24 @@ def elu(
     return refs.where(refs.gt(a, 0), a, rhs)
 
 
+@register_decomposition(torch.ops.aten.relu)
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a",),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def relu(
+    a: TensorLikeType, inplace: bool = False
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.relu
+    """
+
+    if inplace:
+        raise NotImplementedError
+
+    return torch.where(torch.gt(a, 0), a, torch.zeros_like(a))
```

Comment:
```suggestion
    return torch.where(torch.gt(a, 0), a, 0)
```
to avoid materializing full size 0 tensor.

Reply:
This doesn't propagate `nans` like eager does:
```
In [1]: import torch

In [2]: a=torch.tensor([float('nan'), -1., 1., 0.])

In [3]: torch.relu(a)
Out[3]: tensor([nan, 0., 1., 0.])

In [4]: torch.where(torch.gt(a, 0), a, 0)
Out[4]: tensor([0., 0., 1., 0.])
```
On one hand, I have been on record saying "activations should use `where` where appropriate, it makes more sense and leads to more sensible gradients where people have no reason to ask why aren't you splitting grad when values are equal, like minimum and maximum do". On another, this breaks with existing tradition of aggressively propagating nans, and we are definitely biasing towards existing nan propagation behavior.

Reply:
done